nifi-nar-bundles/nifi-standard-services/nifi-record-serialization-services-bundle/nifi-record-serialization-services/src/test/java/org/apache/nifi/xml/TestXMLReader.java

```diff
@@ -31,80 +33,99 @@
-
+    private TestRunner setup(Map<PropertyDescriptor, String> xmlReaderProperties) throws InitializationException {
         TestRunner runner = TestRunners.newTestRunner(TestXMLReaderProcessor.class);
-        reader = new XMLReader();
+        XMLReader reader = new XMLReader();
+
         runner.addControllerService("xml_reader", reader);
         runner.setProperty(TestXMLReaderProcessor.XML_READER, "xml_reader");
 
-        final String outputSchemaText = new String(Files.readAllBytes(Paths.get(filePath)));
-        runner.setProperty(reader, SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_TEXT_PROPERTY);
-        runner.setProperty(reader, SchemaAccessUtils.SCHEMA_TEXT, outputSchemaText);
+        for (Map.Entry<PropertyDescriptor, String> entry : xmlReaderProperties.entrySet()) {
+            runner.setProperty(reader, entry.getKey(), entry.getValue());
+        }
 
+        runner.enableControllerService(reader);
         return runner;
     }
 
     @Test
-    public void testRecordFormat() throws IOException, InitializationException {
-        TestRunner runner = setup("src/test/resources/xml/testschema");
-
-        runner.setProperty(reader, XMLReader.RECORD_FORMAT, XMLReader.RECORD_EVALUATE);
-
-        runner.enableControllerService(reader);
-
+    public void testRecordFormatDeterminedBasedOnAttribute() throws IOException, InitializationException {
+        // GIVEN
+        String outputSchemaPath = "src/test/resources/xml/testschema";
+        String outputSchemaText = new String(Files.readAllBytes(Paths.get(outputSchemaPath)));
+
+        Map<PropertyDescriptor, String> xmlReaderProperties = new HashMap<>();
+        xmlReaderProperties.put(SchemaAccessUtils.SCHEMA_ACCESS_STRATEGY, SchemaAccessUtils.SCHEMA_TEXT_PROPERTY.getValue());
+        xmlReaderProperties.put(SchemaAccessUtils.SCHEMA_TEXT, outputSchemaText);
+        xmlReaderProperties.put(XMLReader.RECORD_FORMAT, XMLReader.RECORD_EVALUATE.getValue());
+        TestRunner runner = setup(xmlReaderProperties);
+
+        // WHEN
         InputStream is = new FileInputStream("src/test/resources/xml/people.xml");
         runner.enqueue(is, Collections.singletonMap(EVALUATE_IS_ARRAY, "true"));
         runner.run();
 
+        // THEN
         List<MockFlowFile> flowFile = runner.getFlowFilesForRelationship(TestXMLReaderProcessor.SUCCESS);
         List<String> records = Arrays.asList((new String(runner.getContentAsByteArray(flowFile.get(0)))).split("\n"));
 
         assertEquals(4, records.size());
     }
 
     @Test
-    public void testRecordFormat2() throws IOException, InitializationException {
-        TestRunner runner = setup("src/test/resources/xml/testschema");
-
-        runner.setProperty(reader, XMLReader.RECORD_FORMAT, XMLReader.RECORD_ARRAY);
-
-        runner.enableControllerService(reader);
-
+    public void testRecordFormatArray() throws IOException, InitializationException {

```

Comment:
Using the following input: 

`<record>
  <num>123</num>
  <software favorite="true">Apache NiFi</software>
</record> `

and the value "body" for "content field" property,  With JsonRecordSetWriter I got the following result:

`[{"num":123,"software":{"favorite":true,"body":"Apache NiFi"},"favorite":null}]`

Is the favorite with the null value expected?

Reply:
No, that is a bug fixed by NIFI-9831.